New model reg revamp (#523)
wandb/docs: docs/guides/model_registry/model-management-concepts.md

---
description: 'Model Registry terms and concepts'
displayed_sidebar: default
---

# Terms and concepts

<head>
  <title>Model Registry terms and concepts</title>
</head>

The following terms are used to describe the W&B Model Registry: [*model version*](#model-version), [*model artifact*](#model-artifact), and [*registered model*](#registered-model).

## Model version
A model version is an immutable directory of data and metadata that describes a trained model. Add files that let you store (and restore) your model architecture and learned parameters.

Store files within model versions that are produced from the serialization process provided by your modeling library (for example, [PyTorch](https://pytorch.org/tutorials/beginner/saving\_loading\_models.html) and [Keras](https://www.tensorflow.org/guide/keras/save\_and\_serialize)).

> A model version belongs to one, and only one, [model artifact](#model-artifact). A model version can belong to zero or more, [registered models](#registered-model).

Reviewer: Maybe we can make a note here highlighting that a model version represents a single model checkpoint - its a snapshot at a point in time of a model and its files within an experiment

<!-- [INSERT IMAGE] -->

## Model alias

> [INSERT - to do]

Author: @ibindlish Can you provide a definition for model aliases. :)?

Reviewer 2: Model aliases are strings that allow you to uniquely identify each version of your registered model. You can't assign the same alias to multiple versions of a registered model.

Reviewer: Lets reframe to be in the affirmative:

> Model aliases are mutable, strings that allow you to uniquely identify or reference each version of your registered model used a semantically-related keyword. You can only assign an alias to one version of a registered model. This is because an alias should refer to a unique version when used programatically. It also allows aliases to be used to capture a model's state (champion, candidate, production). 

TBH I think integrating specific example of what best practice aliases are such as MLFlow does works super well
<img width="845" alt="Screenshot 2024-01-08 at 5 30 57 PM" src="https://github.com/wandb/docodile/assets/40642416/330ef911-e9a8-4cac-a48b-dcde358152fe">

It is common practice to use aliases such as  "best", "latest", "production", or "staging" to mark model versions with special purposes.

## Model artifact

> A model artifact is a sequence of logged model versions. 

Reviewer: We should be specific with what we mean w sequence here I think - i.e logged in order changes were made. 

We can also add a note - a model artifact might be empty if no model versions have been logged to it, or contain 1+ model versions.

<!-- Model artifacts can alias specific versions so that downstream consumers can pin that alias.  -->

For example, suppose you create a model artifact. During model training, you create model a version when you periodically save checkpoints. Each checkpoint will correspond to its own [model version](#model-version). All of the model versions will belong to the model artifact you created at the beginning of your training script.

The proceeding image shows a model artifact that contains three model versions: v0, v1, and v2.

![](@site/static/images/models/mr1c.png)

View an [example model artifact](https://wandb.ai/timssweeney/model\_management\_docs\_official\_v0/artifacts/model/mnist-zws7gt0n).

:::tip
Model artifacts often track the progress of training.
:::

## Registered model
A registered model is a collection of pointers (links) to model versions. You can think of a registered model as a folder of "bookmarks". Each "bookmark" of a registered model is a pointer to a [model version](#model-version) that belongs to a [model artifact](#model-artifact). 

> Registered models often represent candidate models for a single modeling use case or task. For example, [INSERT EXAMPLE].

Author: Anyone have a superb example they'd like to share?

Reviewer: Sure, here are some from my friend chatgpt

Image Classification:

"ImageClassifier-ResNet50"
"ImageClassifier-VGG16"
"DogBreedClassifier-MobileNetV2"

Object Detection:

"ObjectDetector-YOLOv3"
"ObjectDetector-FasterRCNN"

Sentiment Analysis:

"SentimentAnalyzer-LSTM"
"SentimentClassifier-BERT"
Named Entity Recognition (NER):

"NERModel-BiLSTM-CRF"
"NERClassifier-BERT"

Language Translation:

"Translator-Transformer"
"LanguageModel-Seq2Seq"

Time Series Forecasting:

"TimeSeriesForecaster-LSTM"
"ForecastingModel-Prophet"

Generative Models:

"ImageGenerator-DCGAN"
"FashionGenerator-StyleGAN2

> View an [example Registered Model](https://wandb.ai/timssweeney/model\_management\_docs\_official\_v0/artifacts/model/MNIST%20Grayscale%2028x28).

Reviewer: lets use a more up to date example in our demo repo - https://wandb.ai/reviewco/registry/model?selectionPath=reviewco%2Fmodel-registry%2FFinetuned-Review-Autocompletion&view=versions

</head>

The following terms describe key components of the W&B Model Registry: [*model version*](#model-version), [*model artifact*](#model-artifact), and [*registered model*](#registered-model).

## Model version
A model version represents a single model checkpoint. Model versions are a snapshot at a point in time of a model and its files within an experiment. 

A model version is an immutable directory of data and metadata that describes a trained model. Add files that let you store (and restore) your model architecture and learned parameters. 

A model version belongs to one, and only one, [model artifact](#model-artifact). A model version can belong to zero or more, [registered models](#registered-model). Model versions are stored in a model artifact in the order they are logged to the model artifact. 

Store files within model versions that are produced from the serialization process provided by your modeling library (for example, [PyTorch](https://pytorch.org/tutorials/beginner/saving\_loading\_models.html) and [Keras](https://www.tensorflow.org/guide/keras/save\_and\_serialize)).

<!-- [INSERT IMAGE] -->

## Model alias

> Model aliases are mutable strings that allow you to uniquely identify or reference each version of your registered model used a semantically-related keyword. You can only assign an alias to one version of a registered model. This is because an alias should refer to a unique version when used programatically. It also allows aliases to be used to capture a model's state (champion, candidate, production).

Reviewer: ```suggestion
Model aliases are mutable strings that allow you to uniquely identify or reference each version of your registered model used a semantically-related identifier. You can only assign an alias to one version of a registered model. This is because an alias should refer to a unique version when used programatically. It also allows aliases to be used to capture a model's state (champion, candidate, production). Note, these are distinct from tags which are added to registered models, which are akin to keywords or labels and can be added to more than 1 registered models. 
```

Author: Let's also add a new definition here called "Model tags"?

Reviewer: Yes I think this is a great idea - and we can link the chapter on model tags to learn more about how to use them. 

Model Tags: Model Tags are used to organize registered models into categories and to search over those categories in the Model Registry's search bar. They appear at the top of the Registered Model Card. You might choose to use them to group your registered models by ML task, owning team, or priority. The same tag can be added to multiple registered models to allow for grouping. Please note that tags, which are labels applied to registered models for grouping and discoverability, are different from model aliases (defined above) which serve as unique identifiers or nicknames to fetch one model version programatically. To learn more about using tags to organize the tasks in your Model Registry, see [this](https://docs.wandb.ai/guides/model_registry/model_tags) page.

Reviewer: ^ left a sample definition above

It is common practice to use aliases such as  "best", "latest", "production", or "staging" to mark model versions with special purposes.

For example, suppose you create a model and assign it a `"best"` alias. You can refer to that specific model with `run.use_model` 

```python
import wandb
run = wandb.init()
name = f"{entity/project/model_artifact_name}:{alias}"
run.use_model(name=name)
```

## Model artifact

> A model artifact is a group of logged [model versions](#model-version). Model versions are stored in a model artifact in the order they are logged to the model artifact. 

Reviewer: ```suggestion
A model artifact is a collection of logged [model versions](#model-version). Model versions are stored in a model artifact in the order they are logged to the model artifact. 
```

A model artifact can contain one or more model versions. A model artifact can be empty if no model versions are logged to it. 

> For example, suppose you create a model artifact. During model training, you create model a version when you periodically save checkpoints. Each checkpoint will correspond to its own [model version](#model-version). All of the model versions belong to the model artifact you created at the beginning of your training script.

Reviewer: ```suggestion
For example, suppose you create a model artifact. During model training, you create model a version when you periodically save checkpoints. Each checkpoint will correspond to its own [model version](#model-version). All of the model versions belong to the model artifact you created at the beginning of your training script and will be assigned a version number depending on the sequence in which they were logged. A new version is automatically created when the contents of the latest version that was logged has changed. 
```

The proceeding image shows a model artifact that contains three model versions: v0, v1, and v2.

![](@site/static/images/models/mr1c.png)

View an [example model artifact here](https://wandb.ai/timssweeney/model\_management\_docs\_official\_v0/artifacts/model/mnist-zws7gt0n).

## Registered model

> A registered model is a collection of pointers (links) to model versions. You can think of a registered model as a folder of "bookmarks". Each "bookmark" of a registered model is a pointer to a [model version](#model-version) that belongs to a [model artifact](#model-artifact). 

Reviewer: ```suggestion
A registered model is a collection of pointers (links) to model versions. You can think of a registered model as a folder of "bookmarks" of candidate models for the same ML Task. Each "bookmark" of a registered model is a pointer to a [model version](#model-version) that belongs to a [model artifact](#model-artifact). Model versions are assigned version numbers in the order in which they were linked to the registered model. 
```

> A registered model is a collection of pointers (links) to model versions. You can think of a registered model as a folder of "bookmarks". Each "bookmark" of a registered model is a pointer to a [model version](#model-version) that belongs to a [model artifact](#model-artifact). 

Reviewer: ```suggestion
A registered model is a collection of pointers (links) to model versions. You can think of a registered model as a folder of "bookmarks". Each "bookmark" of a registered model is a pointer to a [model version](#model-version) that belongs to a [model artifact](#model-artifact). You can use [tags](https://docs.wandb.ai/guides/model_registry/model_tags) to group your registered models. 
```
